samples/bluetooth/direct_test_mode/src/dtm.c

```diff
@@ -1683,94 +1744,111 @@ static uint32_t constant_tone_setup(uint8_t cte_info)
-		dtm_inst.txpower = dtm_hw_radio_min_power_get();
-		dtm_inst.event =
-			((dtm_inst.txpower <<
-			  LE_TRANSMIT_POWER_RESPONSE_LVL_POS) &
-			 LE_TRANSMIT_POWER_RESPONSE_LVL_MASK) |
-			LE_TRANSMIT_POWER_MIN_LVL_BIT;
+	return fem_tx_output_power_min_get(frequency);
+}
 
-		return DTM_SUCCESS;
-	}
+static int8_t dtm_radio_max_power_get(uint16_t frequency)
+{
+	return fem_tx_output_power_max_get(frequency);
+}
 
-	if (parameter == LE_TRANSMIT_POWER_LVL_SET_MAX) {
-		dtm_inst.txpower = dtm_hw_radio_max_power_get();
-		dtm_inst.event =
-			((dtm_inst.txpower <<
-			  LE_TRANSMIT_POWER_RESPONSE_LVL_POS) &
-			 LE_TRANSMIT_POWER_RESPONSE_LVL_MASK) |
-			LE_TRANSMIT_POWER_MAX_LVL_BIT;
+static int8_t dtm_radio_nearest_power_get(int8_t tx_power, uint16_t frequency)
+{
+	int8_t tx_power_floor = fem_tx_output_power_check(tx_power, frequency, false);
+	int8_t tx_power_ceiling = fem_tx_output_power_check(tx_power, frequency, true);
+	int8_t output_power;
 
-		return DTM_SUCCESS;
-	}
+	output_power = (abs(tx_power_floor - tx_power) > abs(tx_power_ceiling - tx_power)) ?
+		       tx_power_ceiling : tx_power_floor;
 
-	if ((parameter < LE_TRANSMIT_POWER_LVL_MIN) ||
-	    (parameter > LE_TRANSMIT_POWER_LVL_MAX)) {
-		dtm_inst.event =
-			((dtm_inst.txpower <<
-			  LE_TRANSMIT_POWER_RESPONSE_LVL_POS) &
-			 LE_TRANSMIT_POWER_RESPONSE_LVL_MASK) |
-			LE_TEST_STATUS_EVENT_ERROR;
+	return output_power;
+}
 
-		if (dtm_inst.txpower == dtm_hw_radio_min_power_get()) {
-			dtm_inst.event |= LE_TRANSMIT_POWER_MIN_LVL_BIT;
-		} else if (dtm_inst.txpower == dtm_hw_radio_max_power_get()) {
-			dtm_inst.event |= LE_TRANSMIT_POWER_MAX_LVL_BIT;
-		} else {
-			/* Do nothing. */
-		}
+#else
+static int8_t dtm_radio_min_power_get(uint16_t frequency)
+{
+	ARG_UNUSED(frequency);
 
-		return DTM_ERROR_ILLEGAL_CONFIGURATION;
-	}
+	return dtm_hw_radio_min_power_get();
+}
 
-	if (parameter <= ((int8_t) dtm_hw_radio_min_power_get())) {
-		dtm_inst.txpower = dtm_hw_radio_min_power_get();
-		dtm_inst.event =
-			((dtm_inst.txpower <<
-			  LE_TRANSMIT_POWER_RESPONSE_LVL_POS) &
-			 LE_TRANSMIT_POWER_RESPONSE_LVL_MASK) |
-			LE_TRANSMIT_POWER_MIN_LVL_BIT;
+static int8_t dtm_radio_max_power_get(uint16_t frequency)
+{
+	ARG_UNUSED(frequency);
 
-		return DTM_SUCCESS;
-	}
+	return dtm_hw_radio_max_power_get();
+}
 
-	if (parameter >= ((int8_t) dtm_hw_radio_max_power_get())) {
-		dtm_inst.txpower = dtm_hw_radio_max_power_get();
-		dtm_inst.event =
-			((dtm_inst.txpower <<
-			  LE_TRANSMIT_POWER_RESPONSE_LVL_POS) &
-			 LE_TRANSMIT_POWER_RESPONSE_LVL_MASK) |
-			LE_TRANSMIT_POWER_MAX_LVL_BIT;
+static int8_t dtm_radio_nearest_power_get(int8_t tx_power, uint16_t frequency)
+{
+	int8_t output_power = INT8_MAX;
+	const size_t size = dtm_hw_radio_power_array_size_get();
+	const uint32_t *power = dtm_hw_radio_power_array_get();
 
-		return DTM_SUCCESS;
-	}
+	ARG_UNUSED(frequency);
 
-	/* Look for the nearest tansmit power level and set it. */
 	for (size_t i = 1; i < size; i++) {
```

Comment:
Actually, you could use `mpsl_tx_power_radio_supported_power_adjust` instead and use a similar approach as in automatic control mode. `mpsl_tx_power_radio_supported_power_adjust` is in the MPSL public API. However it would require calling `mpsl_fem_init` at some point earliear in time (as the function for initializing the tx_power API is not available in the public API - `mpsl_fem_init` wouldn't do any harm even when no FEM is present).

This would also remove the need of the existence of the `nrf_power_value` array, as it could be replaced by the code in the MPSL library.

However this is just a suggestion, as the code here seems a bit complicated.

Reply:
I would be good to use this function here. However MPSL library is not linked into build when bulding for targets without front-end modules